presto-main/src/main/java/com/facebook/presto/sql/planner/DesugarRowSubscriptRewriter.java

```diff
@@ -0,0 +1,125 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.sql.planner;
+
+import com.facebook.presto.Session;
+import com.facebook.presto.metadata.Metadata;
+import com.facebook.presto.spi.type.RowType;
+import com.facebook.presto.spi.type.Type;
+import com.facebook.presto.sql.parser.SqlParser;
+import com.facebook.presto.sql.tree.Cast;
+import com.facebook.presto.sql.tree.DereferenceExpression;
+import com.facebook.presto.sql.tree.Expression;
+import com.facebook.presto.sql.tree.ExpressionRewriter;
+import com.facebook.presto.sql.tree.ExpressionTreeRewriter;
+import com.facebook.presto.sql.tree.Identifier;
+import com.facebook.presto.sql.tree.LongLiteral;
+import com.facebook.presto.sql.tree.NodeRef;
+import com.facebook.presto.sql.tree.SubscriptExpression;
+import com.facebook.presto.sql.tree.SymbolReference;
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.ImmutableMap;
+
+import java.util.Map;
+import java.util.Optional;
+
+import static java.lang.Math.toIntExact;
+import static java.util.Objects.requireNonNull;
+
+/**
+ * Replaces subscript expression on Row with cast and dereference:
+ *
+ * <pre>
+ * ROW(1, 'a', 2)[2]
+ * </pre>
+ * <p>
+ * is transformed into:
+ *
+ * <pre>
+ *     (CAST (ROW (1, 'a', 2) AS ROW (field_0 bigint, field_1 varchar(1), field_2 bigint))).field_1
+ * </pre>
+ */
+public class DesugarRowSubscriptRewriter
+{
+    private DesugarRowSubscriptRewriter()
+    {
+    }
+
+    public static Expression rewrite(Expression expression, Map<NodeRef<Expression>, Type> expressionTypes)
+    {
+        return ExpressionTreeRewriter.rewriteWith(new Visitor(expressionTypes), expression, null);
+    }
+
+    public static Expression rewrite(Expression expression, Session session, Metadata metadata, SqlParser sqlParser, SymbolAllocator symbolAllocator)
+    {
+        requireNonNull(metadata, "metadata is null");
+        requireNonNull(sqlParser, "sqlParser is null");
+
+        if (expression instanceof SymbolReference) {
+            return expression;
+        }
+        return new AnalyzedExpressionRewriter(session, metadata, sqlParser, symbolAllocator.getTypes()).rewriteWith(Visitor::new, expression);
+    }
+
+    private static class Visitor
+            extends ExpressionRewriter<Void>
+    {
+        private final Map<NodeRef<Expression>, Type> expressionTypes;
+
+        public Visitor(Map<NodeRef<Expression>, Type> expressionTypes)
+        {
+            this.expressionTypes = ImmutableMap.copyOf(requireNonNull(expressionTypes, "expressionTypes is null"));
+        }
+
+        @Override
+        public Expression rewriteSubscriptExpression(SubscriptExpression node, Void context,
```

Comment:
ditto, for this case I would say merge the next line, since it's not really long